File: rest/src/main/java/com/droidtitan/volleyexamples/rest/module/RestModule.java
package com.droidtitan.volleyexamples.rest.module;

import android.content.Context;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.ImageLoader;
import com.android.volley.toolbox.Volley;
import com.droidtitan.volleyexamples.rest.fragment.ImageLoaderFragment;
import com.droidtitan.volleyexamples.rest.fragment.NetworkImageFragment;
import com.droidtitan.volleyexamples.rest.fragment.RequestFragment;
import com.droidtitan.volleyexamples.rest.util.LruBitmapCache;
import com.droidtitan.volleyexamples.rest.util.OkHttpStack;
import com.squareup.okhttp.OkHttpClient;
import dagger.Module;
import dagger.Provides;

import javax.inject.Named;
import javax.inject.Singleton;

@Module(injects = {RequestFragment.class, NetworkImageFragment.class, ImageLoaderFragment.class},
        library = true, complete = false)
public class RestModule {

    private Context context;

    public RestModule(Context appContext) {
        context = appContext;
    }

    @Provides
    @Named("App")
    public Context provideAppContext() {
        return context;
    }

    @Singleton
    @Provides
    public OkHttpClient provideOkHttpClient() {
        return new OkHttpClient();
    }

    @Provides
    @Singleton
    public RequestQueue provideRequestQueue(OkHttpClient okHttpClient, @Named("App") Context context) {
        /** Set up to use OkHttp */
        return Volley.newRequestQueue(context, new OkHttpStack(okHttpClient));
    }

    @Provides
    @Singleton
    public ImageLoader provideImageLoader(RequestQueue requestQueue) {
        final int imgSize = 1024;
        final int count = 8;
        int maxSize = (int) (Runtime.getRuntime().maxMemory() / imgSize / count);
        return new ImageLoader(requestQueue, new LruBitmapCache(maxSize));
    }
}

File: app/src/main/java/com/droidtitan/volley/di/RestModule.kt
package com.droidtitan.volley.di

import android.app.Application
import android.content.Context
import android.graphics.Bitmap
import android.support.v4.util.LruCache
import com.android.volley.RequestQueue
import com.android.volley.toolbox.HurlStack
import com.android.volley.toolbox.ImageLoader
import com.android.volley.toolbox.ImageLoader.ImageCache
import com.android.volley.toolbox.Volley
import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.OkUrlFactory
import dagger.Module
import dagger.Provides
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import javax.inject.Singleton

@Module
public class RestModule(val application: Application) {

    @Singleton @Provides
    public fun provideQueue(): RequestQueue {

        val factory = OkUrlFactory(OkHttpClient())
        val hurlStack = object : HurlStack() {
            @Throws(IOException::class) override fun createConnection(url: URL): HttpURLConnection {
                return factory.open(url)
            }
        }

        return Volley.newRequestQueue(application, hurlStack)
    }

    @Singleton @Provides
    public fun provideImageLoader(queue: RequestQueue): ImageLoader {

        val imageSize = 1024L
        val count = 8L
        val maxSize = (Runtime.getRuntime().maxMemory() / imageSize / count).toInt()

        val lruCache = object : LruCache<String, Bitmap>(maxSize), ImageCache {
            override fun sizeOf(key: String?, value: Bitmap?) = value!!.rowBytes * value.height
            override fun getBitmap(url: String): Bitmap? = get(url)
            override fun putBitmap(url: String, bitmap: Bitmap) = put(url, bitmap).let {  }
        }

        return ImageLoader(queue, lruCache)
    }
}